javaslang/src/main/java/javaslang/collection/AbstractSet.java

```diff
@@ -0,0 +1,49 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import javaslang.Value;
+
+import java.util.function.Predicate;
+
+/**
+ * @since 2.0.5
+ */
+public abstract class AbstractSet<T> implements Set<T> {
+
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public boolean equals(Object obj) {
+        return trueIfSameOrWhenInstanceOfASet(obj, this::euqalsTypeUnsafe);
+    }
+
+    @Override
+    public boolean eq(Object obj) {
```

Comment:
Overwriting `eq` would break the contract of structural equality. E.g. this following must be true:

```java
TreeSet.of(1, 2).eq(Arrays.asList(1, 2)) == true

TreeSet.of(1, 2).eq(Arrays.asList(2, 1)) == false
```